File: dashr/chapters/graph-crossfiltering/examples/interactive-graph-2.R
library(dashR)
library(dashCoreComponents)
library(dashHtmlComponents)

app <- Dash$new()

df <- read.csv('dashr/chapters/callbacks/examples/indicators.csv', header = TRUE, sep = ",")
available_indicators <- unique(df$Indicator_Name)
option_indicator <- list()
for (i in 1:length(available_indicators)){
  option_indicator[[i]] <- list(label = available_indicators[i], value = available_indicators[i])
}


app$layout_set(
  htmlDiv(list(
    htmlDiv(list(
      htmlDiv(list(
        dccDropdown(
          id = 'crossfilter-xaxis-column',
          options = option_indicator,
          value = 'Fertility rate, total (births per woman)'
        ),
        dccRadioItems(
          id = 'crossfilter-xaxis-type',
          options = list(list(label = 'Linear', value = 'linear'),
                         list(label = 'Log', value = 'log')),
          value = 'linear',
          labelStyle = list(display = 'inline-block')
        )
      ), style = list(width = '49%', display = 'inline-block')),

      htmlDiv(list(
        dccDropdown(
          id = 'crossfilter-yaxis-column',
          options = option_indicator,
          value = 'Life expectancy at birth, total (years)'
        ),
        dccRadioItems(
          id = 'crossfilter-yaxis-type',
          options = list(list(label = 'Linear', value = 'linear'),
                         list(label = 'Log', value = 'log')),
          value = 'linear',
          labelStyle = list(display = 'inline-block')
        )
      ), style = list(width = '49%', flaot = 'display', display = 'inline-block'))
    ), style = list(
      borderBottom = 'thin lightgrey solid',
      backgroundColor = 'rgb(250, 250, 250)',
      padding = '10px 5px')
    ),

    htmlDiv(list(
      dccGraph(
        id = 'crossfilter-indicator-scatter',
        hoverData = list(points = list(list('customdata' = 'Japan')))
      )), style = list(
        width ='49%',
        display = 'inline-block',
        padding = '0 20')
    ),

    htmlDiv(list(
      dccGraph(id='x-time-series'),
      dccGraph(id='y-time-series')
    ), style=list(display = 'inline-block', width = '49%')),

    #dccGraph(id = 'indicator-graphic'),
    htmlDiv(list(
      dccSlider(
        id = 'crossfilter-year--slider',
        min = 1,
        max = length(unique(df$Year)),
        marks = unique(df$Year),
        value = length(unique(df$Year))
      )
    ), style = list(width = '49%', padding = '0px 20px 20px 20px'))
  ))
)

create_time_series <- function(dff, axis_type, title){
  print(dff)
  return(list(
    'data' = list(list(
      x = dff[['Year']],
      y = dff[['Value']],
      mode = 'lines+markers'
    )),
    'layout' = list(
      height = 225,
      margin = list('l' = 20, 'b' = 30, 'r' = 10, 't' = 10),
      'annotations' = list(list(
        x = 0, 'y' = 0.85, xanchor = 'left', yanchor = 'bottom',
        xref = 'paper', yref = 'paper', showarrow = FALSE,
        align = 'left', bgcolor = 'rgba(255, 255, 255, 0.5)',
        text = 'linear'
      )),
      yaxis = list(type = axis_type),
      xaxis = list(showgrid = FALSE)
    )
  ))
}

app$callback(
  output = list(id='crossfilter-indicator-scatter', property='figure'),
  params = list(input(id='crossfilter-xaxis-column', property='value'),
                input(id='crossfilter-yaxis-column', property='value'),
                input(id='crossfilter-xaxis-type', property='value'),
                input(id='crossfilter-yaxis-type', property='value'),
                input(id='crossfilter-year--slider', property='value')),
  function(xaxis_column_name, yaxis_column_name, xaxis_type, yaxis_type, year_value) {
    selected_year <- unique(df$Year)[year_value]
    dff <- selected_year
    traces <- list()
    if (selected_year %in% unique(df$Year)){
      filtered_df <- df[df[["Year"]] %in% selected_year, ]
      traces[[1]] <- list(
        x = filtered_df[filtered_df$Indicator_Name %in% xaxis_column_name, "Value", drop = TRUE],
        y = filtered_df[filtered_df$Indicator_Name %in% yaxis_column_name, "Value", drop = TRUE],
        opacity=0.7,
        text = filtered_df[filtered_df$Indicator_Name %in% yaxis_column_name, "Country_Name", drop = TRUE],
        mode = 'markers',
        marker = list(
          'size'= 15,
          'opacity' = 0.5,
          'line' = list('width' = 0.5, 'color' = 'white')
        )
      )

      return (list(
        'data' = traces,
        'layout'= list(
          xaxis = list('title' = xaxis_column_name, 'type' = xaxis_type),
          yaxis = list('title' = yaxis_column_name, 'type' = yaxis_type),
          margin = list('l' = 40, 'b' = 30, 't' = 10, 'r' = 0),
          height = 450,
          hovermode = 'closest'
        )
      ))
    }
  }
)

app$callback(
  output = list(id='x-time-series', property='figure'),
  params = list(input(id='crossfilter-indicator-scatter', property='hoverData'),
                input(id='crossfilter-xaxis-column', property='value'),
                input(id='crossfilter-xaxis-type', property='value')),
  function(hoverData, xaxis_column_name, axis_type) {
    #print(hoverData)
    #print("Nadia")
    country_name = hoverData$points[[1]]$customdata
    #print(country_name)
    #print("Nadia")
    dff <- splitdf[df$Country_Name %in% country_name]
    #print(df)
    #print("Nadia")
    dff <- df[df$Indicator_Name %in% xaxis_column_name]
    title = paste(c(country_name, xaxis_column_name), sep = '<b>')
    #browser()
    return(create_time_series(dff, axis_type, title))
  }
)


app$run_heroku()

File: dash_docs/chapters/graph_crossfiltering/examples/interactive-graph-2.jl
using CSV, DataFrames, Dash, DashHtmlComponents, DashCoreComponents
using PlotlyJS


url = "https://plotly.github.io/datasets/country_indicators.csv"
download(url, "country-indicators.csv")
df = DataFrame(CSV.File("country-indicators.csv"))

dropmissing!(df)

rename!(df, Dict(:"Indicator Name" => "Indicator"))
rename!(df, Dict(:"Country Name" => "Country"))

available_indicators = unique(df[:, "Indicator"])
years = unique(df[:, "Year"])

app = dash()

app.layout = html_div() do
    html_div(
        children = [
            html_div(
                children = [
                    dcc_dropdown(
                        id = "crossfilter-xaxis-column",
                        options = [
                            (label = i, value = i)
                            for i in available_indicators
                        ],
                        value = "Fertility rate, total (births per woman)",
                    ),
                    dcc_radioitems(
                        id = "crossfilter-xaxis-type",
                        options = [
                            (label = i, value = i) for i in ["linear", "log"]
                        ],
                        value = "linear",
                    ),
                ],
                style = (width = "49%", display = "inline-block"),
            ),
            html_div(
                children = [
                    dcc_dropdown(
                        id = "crossfilter-yaxis-column",
                        options = [
                            (label = i, value = i)
                            for i in available_indicators
                        ],
                        value = "Life expectancy at birth, total (years)",
                    ),
                    dcc_radioitems(
                        id = "crossfilter-yaxis-type",
                        options = [
                            (label = i, value = i) for i in ["linear", "log"]
                        ],
                        value = "linear",
                    ),
                ],
                style = (
                    width = "49%",
                    float = "right",
                    display = "inline-block",
                ),
            ),
        ],
        style = (
            borderBottom = "thin lightgrey solid",
            backgroundColor = "rgb(250, 250, 250)",
            padding = "10px 5px",
        ),
    ),
    html_div(
        children = [dcc_graph(id = "crossfilter-indicator-scatter"),
                    dcc_slider(
                        id = "crossfilter-year-slider",
                        min = 0,
                        max = length(years) - 1,
                        value = 0,
                        marks = years,
                    )],
        style = (width = "49%", display = "inline-block", paddingBottom = "400px"),
    ),
    html_div(
        children = [
            dcc_graph(id = "x-time-series"),
            dcc_graph(id = "y-time-series"),
        ],
        style = (width = "49%", display = "inline-block"),
    )
end

callback!(
    app,
    Output("crossfilter-indicator-scatter", "figure"),
    Input("crossfilter-xaxis-column", "value"),
    Input("crossfilter-yaxis-column", "value"),
    Input("crossfilter-xaxis-type", "value"),
    Input("crossfilter-yaxis-type", "value"),
    Input("crossfilter-year-slider", "value"),
) do x_indicator, y_indicator, x_axis_type, y_axis_type, year_slider_value

    dff = df[df.Year .== years[year_slider_value+1], :]

    x_indicator_data = dff[dff.Indicator .== x_indicator, :][:, :Value]
    y_indicator_data = dff[dff.Indicator .== y_indicator, :][:, :Value]
    country_names = dff[dff.Indicator .== y_indicator, :][:, :Country]

    figure = (
        data = [
            (
                x = x_indicator_data,
                y = y_indicator_data,
                type = "scatter",
                mode = "markers",
                text = country_names
            ),
        ],
        layout = (
            xaxis = ((title = x_indicator), (type = x_axis_type)),
            yaxis = ((title = y_indicator), (type = y_axis_type)),
        ),
    )

    return figure

end

callback!(
    app,
    Output("x-time-series", "figure"),
    Input("crossfilter-indicator-scatter", "hoverData"),
    Input("crossfilter-xaxis-column", "value"),
    Input("crossfilter-xaxis-type", "value"),
) do hoverData, x_indicator, axis_type

    if hoverData == nothing
        return PreventUpdate()
    end

    country = hoverData[1][1][:text]

    dff = df[df.Country .== country, :]
    dff = dff[dff.Indicator .== x_indicator, :]

    figure = (
        data = [
            (
                x = dff[!, "Year"],
                y = dff[!, "Value"],
                type = "scatter",
                mode = "markers+lines"
            )
        ],
        layout = (
            title = country,
            xaxis = ((title = "Year"), (type = axis_type)),
            yaxis = ((title = x_indicator), (type = axis_type)),
        ),
    )

    return figure

end

callback!(
    app,
    Output("y-time-series", "figure"),
    Input("crossfilter-indicator-scatter", "hoverData"),
    Input("crossfilter-yaxis-column", "value"),
    Input("crossfilter-yaxis-type", "value"),
) do hoverData, y_indicator, axis_type

    if hoverData == nothing
        return PreventUpdate()
    end

    country = hoverData[1][1][:text]

    dff = df[df.Country .== country, :]
    dff = dff[dff.Indicator .== y_indicator, :]

    figure = (
        data = [
            (
                x = dff[!, "Year"],
                y = dff[!, "Value"],
                type = "scatter",
                mode = "markers+lines"
            )
        ],
        layout = (
            title = country,
            xaxis = ((title = "Year"), (type = axis_type)),
            yaxis = ((title = y_indicator), (type = axis_type)),
        ),
    )

    return figure
end

run_server(app, "0.0.0.0", 8000)
